src/connector/postgres.rs

```diff
@@ -619,18 +620,50 @@ impl TransactionCapable for PostgreSql {}
 impl Queryable for PostgreSql {
     async fn query(&self, q: Query<'_>) -> crate::Result<ResultSet> {
         let (sql, params) = visitor::Postgres::build(q)?;
+
         self.query_raw(sql.as_str(), &params[..]).await
     }
 
     async fn execute(&self, q: Query<'_>) -> crate::Result<u64> {
         let (sql, params) = visitor::Postgres::build(q)?;
+
         self.execute_raw(sql.as_str(), &params[..]).await
     }
 
     #[tracing::instrument(skip(self, params))]
     async fn query_raw(&self, sql: &str, params: &[Value<'_>]) -> crate::Result<ResultSet> {
         metrics::query("postgres.query_raw", sql, params, move || async move {
-            let stmt = self.fetch_cached(sql).await?;
+            let stmt = self.fetch_cached(sql, &[]).await?;
+
+            if stmt.params().len() != params.len() {
+                let kind = ErrorKind::IncorrectNumberOfParameters {
+                    expected: stmt.params().len(),
+                    actual: params.len(),
+                };
+
+                return Err(Error::builder(kind).build());
+            }
+
+            let rows = self
+                .perform_io(self.client.0.query(&stmt, conversion::conv_params(params).as_slice()))
+                .await?;
+
+            let mut result = ResultSet::new(stmt.to_column_names(), Vec::new());
+
+            for row in rows {
+                result.rows.push(row.get_result_row()?);
+            }
+
+            Ok(result)
+        })
+        .await
+    }
+
+    #[tracing::instrument(skip(self, params))]
+    async fn query_raw_typed(&self, sql: &str, params: &[Value<'_>]) -> crate::Result<ResultSet> {
+        metrics::query("postgres.query_raw", sql, params, move || async move {
+            let param_types = conversion::params_to_types(params);
```

Comment:
could we skip this if we already have the statement in the cache? it's a lot of cycles in the worst case...

Reply:
Yeah, that was stupid from me 🤦‍♂️